File: apps/workbench/src/server/routers/monitoring.ts
import { z } from 'zod';
import { router, publicProcedure } from '../trpc';
import { createServicesFromContext } from '../services/ServiceFactory';
import { pythonOCRClient } from '../services/python/PythonOCRClient';
import { pythonConversionClient } from '../services/python/PythonConversionClient';
import { pythonTextClient } from '../services/python/PythonTextClient';

/**
 * Monitoring router for model usage, health, and cost tracking
 */
export const monitoringRouter = router({
  /**
   * Get comprehensive model monitoring data
   */
  getModelMonitoring: publicProcedure.query(async ({ ctx }) => {
    try {
      const { chatService } = createServicesFromContext(ctx);

      // Collect all monitoring data using proper interface methods
      const usage = chatService.getModelUsageStats();
      const capabilities = chatService.getModelCapabilities();
      const health = await chatService.checkAllModelsHealth();

      // Convert capabilities Map to object for JSON serialization
      const capabilitiesObj = capabilities instanceof Map
        ? Object.fromEntries(capabilities)
        : capabilities;

      return {
        usage,
        health,
        capabilities: capabilitiesObj,
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      return {
        usage: [],
        health: [],
        capabilities: {},
        error: error instanceof Error ? error.message : 'Unknown error',
        timestamp: new Date().toISOString(),
      };
    }
  }),

  /**
   * Get model usage statistics only
   */
  getUsageStats: publicProcedure.query(async ({ ctx }) => {
    try {
      const { chatService } = createServicesFromContext(ctx);

      return {
        usage: chatService.getModelUsageStats(),
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      return {
        usage: [],
        error: error instanceof Error ? error.message : 'Unknown error',
        timestamp: new Date().toISOString(),
      };
    }
  }),

  /**
   * Get model health status
   */
  getHealthStatus: publicProcedure.query(async ({ ctx }) => {
    try {
      const { chatService } = createServicesFromContext(ctx);
      const health = await chatService.checkAllModelsHealth();

      return {
        health,
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      return {
        health: [],
        error: error instanceof Error ? error.message : 'Unknown error',
        timestamp: new Date().toISOString(),
      };
    }
  }),

  /**
   * Get model capabilities and pricing
   */
  getCapabilities: publicProcedure.query(async ({ ctx }) => {
    try {
      const { chatService } = createServicesFromContext(ctx);
      const capabilities = chatService.getModelCapabilities();

      // Convert Map to object for JSON serialization
      const capabilitiesObj = capabilities instanceof Map
        ? Object.fromEntries(capabilities)
        : capabilities;

      return {
        capabilities: capabilitiesObj,
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      return {
        capabilities: {},
        error: error instanceof Error ? error.message : 'Unknown error',
        timestamp: new Date().toISOString(),
      };
    }
  }),

  /**
   * Trigger health check for specific model
   */
  checkModelHealth: publicProcedure
    .input(z.object({ modelId: z.string() }))
    .mutation(async ({ ctx, input }) => {
      try {
        const { chatService } = createServicesFromContext(ctx);
        const health = chatService.getModelHealthStatus(input.modelId);

        return {
          health,
          timestamp: new Date().toISOString(),
        };
      } catch (error) {
        return {
          health: null,
          error: error instanceof Error ? error.message : 'Unknown error',
          timestamp: new Date().toISOString(),
        };
      }
    }),

  /**
   * Get Python microservice statistics
   * Shows availability, circuit breaker state, and fallback status
   */
  getPythonServiceStats: publicProcedure.query(() => {
    try {
      const ocrStats = pythonOCRClient.getStats();
      const conversionStats = pythonConversionClient.getStats();
      const textStats = pythonTextClient.getStats();

      return {
        ocr: ocrStats,
        conversion: conversionStats,
        text: textStats,
        summary: {
          allAvailable: ocrStats.available && conversionStats.available && textStats.available,
          anyCircuitOpen:
            ocrStats.circuitBreaker.state === 'OPEN' ||
            conversionStats.circuitBreaker.state === 'OPEN' ||
            textStats.circuitBreaker.state === 'OPEN',
          forceDisabled: ocrStats.forceDisabled || conversionStats.forceDisabled || textStats.forceDisabled,
        },
        timestamp: new Date().toISOString(),
      };
    } catch (error) {
      return {
        error: error instanceof Error ? error.message : 'Unknown error',
        timestamp: new Date().toISOString(),
      };
    }
  }),
});

File: python/artificer_client/resources/monitoring.py
"""
Monitoring resource
"""

from .base import BaseResource
from ..types import ServiceHealth


class Monitoring(BaseResource):
    """Monitoring API resource."""

    def get_health(self) -> ServiceHealth:
        """
        Get service health status.

        Returns:
            Service health information

        Example:
            >>> health = client.monitoring.get_health()
            >>> print(f"Status: {health['status']}")
        """
        return self._trpc_request("monitoring.getHealthStatus")

    def get_python_service_stats(self) -> dict:
        """
        Get Python microservice statistics.

        Shows Python service availability, circuit breaker state, etc.

        Returns:
            Python service statistics

        Example:
            >>> stats = client.monitoring.get_python_service_stats()
            >>> print(f"OCR available: {stats['ocr']['available']}")
            >>> print(f"Circuit state: {stats['ocr']['circuitBreaker']['state']}")
        """
        return self._trpc_request("monitoring.getPythonServiceStats")

    def get_usage_stats(self) -> dict:
        """
        Get model usage statistics.

        Returns:
            Usage statistics

        Example:
            >>> usage = client.monitoring.get_usage_stats()
            >>> for model, stats in usage.items():
            ...     print(f"{model}: {stats['requests']} requests")
        """
        return self._trpc_request("monitoring.getUsageStats")

    def get_model_capabilities(self) -> dict:
        """
        Get available model capabilities and pricing.

        Returns:
            Model capabilities

        Example:
            >>> caps = client.monitoring.get_model_capabilities()
            >>> for model in caps['models']:
            ...     print(f"{model['name']}: ${model['pricing']['input']} per 1M tokens")
        """
        return self._trpc_request("monitoring.getCapabilities")
